Introduce BaseMaster

cofri opened the pull request:
This PR prepares the possibility to support other master algorithms:

- Introduce a `BaseMaster` abstract class with a common API for future master algorithms
- The `DefaultMaster` class is now derived from the `BaseMaster`
- The `Coordinator` is now agnostic to the `DefaultMaster` and creates an instance of Master using `BaseMaster`. Note that this is not yet perfect to support different master algorithms. The Master arguments must be grouped together in a single config dictionary.

The next step is to create a configuration for Master objects, like for stream handlers and storage: the JSON configuration file will gather the Master configuration items (e.g. `cosim_method`, `iterative`, `fixed_point_init`, etc.) in a specific section of the JSON file. The Master subclass will be instantiated based on these config items + the common arguments `fmus`, `connections`, `sequence_order`, etc.

cofri committed:
feat(master): DefaultMaster now inherits from BaseMaster

The DefaultMaster is also registered in the __init__.py file.

cofri committed:
feat(master): introduce BaseMaster

BaseMaster is the abstract base class for master algorithms.
The structure is similar to the one of stream handlers and storages.

Child classes must follow the API with abstract methods such as:
do_step, init_simulation, get_variable, etc.

cofri committed:
feat: Coordinator has no knowledge about the master type

The Coordinator is now agnostic to the master child class. Note that for
now, it is not completely perfect since we still have knowledge of the
arguments to build the Master object, e.g. fixed_point_init.
But this is a first step.

The config parser now builds a config dict for Master inspired from the
stream handler config:
- a "type" key, defining which master child class to instantiate.
- a "config" key, containing all the arguments used to instantiate the
  master algorithm.

M-Mouhcine commented on cofmupy/master/__init__.py:
```diff
@@ -23,6 +23,26 @@
+from .base_master import BaseMaster
 from .default_master import DefaultMaster
+
+# Register all the handlers. This is necessary to be able to create the handlers
```
The comment should refer to BaseMaster subclasses here :)

cofri replied:
Damn copy-paste!

M-Mouhcine commented on cofmupy/config_parser.py:
```diff
@@ -178,34 +178,35 @@ def _build_master_config(self) -> None:
+                        (target.id, target.variable)
+                    )
+
+        self.master_config["type"] = "default"
```
I was wondering if there is a specific reason to separate the `type` from the other configuration parameters. Wouldn't it be better to include the Master type directly in the JSON configuration (under _config_object.master_type), with "default" as the default value?
This would avoid having nested dictionaries (`self.master_config["config"] = config`) that mainly exist to wrap configuration.

cofri replied:
I think we introduce this notation of the dict config in config parser to make a clear distinction between the type of the child class and its configuration parameters (i.e. the kwargs to initialize the child class). We did the same for stream handlers and storage handlers.

This is only done for internal use. In the JSON config file, it's "flattened" as you suggest: the `type` is alongside the other parameters. For example:
```json
"source": {
        "type": "csv",
        "path": "path/to/battery_data.csv",
        "variable": "voltage",
        "unit": "V",
        "interpolation": "previous"
    },
```

We can move `type` inside the config dict to have a single flat dictionary. Both solutions are ok for me.

jfasquel31450 replied:
For me the separation between type and config is useful when we could have different config format pending class type, with the config format as a "black box" from outside of the class. This means that each sub-class is responsible of its config and we can enhance format without impacting neighbor classes.

In the case of the master, I think it's the same, we can keep the separation between type and config if the master config could be different pending master sub-class type.

In any cases, the code below on master initialization should be improved as explain in the "TODO" comment.

cofri replied:
Indeed, the different Master subclasses will have different configuration. For example, our DefaultMaster has `cosim_method` and `iterative`. But OMSimulator does not require these parameters and will need other parameters to configure it. 
The only common parameters are `fmus`, `connections` and `sequence_order` to build the cosimulation system. See my other comment for a suggestion of refactoring.

I don't have a clear opinion on whether the subclass type must be in the config dict or separated.

cofri replied:
After discussion, we've decided to keep the dictionary as is (type + config nested dict), for consistency with other classes in CoFmuPy and to avoid "reserved keys" (like `type`) in the parameters dictionary.

As you suggest, a section in the JSON configuration will be added to configure the master type and its parameters (this will be done in the next PR)

M-Mouhcine reviewed (dismissed):
Good job ! I've only left some minor pendantic remarks.

jfasquel31450 commented on cofmupy/coordinator.py:
```diff
@@ -162,7 +168,10 @@ def start_master(
+        self.master = master_subclass(
             fmu_config_list=config["fmus"],
             connections=config["connections"],
             sequence_order=config["sequence_order"],
```
I agree with the "TODO" comment ; master class should be initialized with only type and config. There is no benefit to breaking down the master configuration, the master class should parse its config itself.

Maybe we could consider DefaultMaster config is now fixed, and make the modification now

cofri replied:
Yes, we could set the DefaultMaster config in the ConfigParser and use `BaseMaster.create_master(master_config)`. But for now the DefaultMaster config is not clearly defined, especially with the `fixed_point_init` and `fixed_point_kwargs` that are parameters of the `Coordinator.start()` and not parameters of the Master config.

Moreover, I think we should think about the way we want to initialize a Master subclass. For now, we have two types of parameters in the `DefaultMaster.__init__`:
- hyper-parameters specific to the Master, e.g. `iterative`, `cosim_method`. These parameters are specific to our DefaultMaster and will not be present in other master subclasses, for instance in OMSimulator.
- parameters that comes from the rest of the config file, like `fmus`, `connections` and `sequence_order`. These are not hyper-parameters of the master, but just information required to build the co-simulation system. These parameters are common to all master subclasses.
I think it will be useful to split these two types of parameters: first ones will be used in the config file for instantiation of the Master object (e.g. `ChildMaster.__init__(cosim_method, iterative, ...)`). The second parameters (`fmus`, `connections`, etc.) must be used elsewhere, for example in a child method `BaseMaster.build_cosimulation(fmus, connections, sequence_order)`.

cofri replied:
From @jfasquel31450 comment, I've moved all constructor arguments for DefaultMaster inside the master_config, in order to instantiate the master object using `BaseMaster.create_master(master_config)`. See commit 61674e2

Note that "fmus" is renamed into "fmu_config_list" to stick to the argument name in DefaultMaster.

jfasquel31450 commented on cofmupy/config_parser.py:
```diff
@@ -178,34 +178,35 @@ def _build_master_config(self) -> None:
             None. Builds self.master_config.
         """
 
+        config = {}
```
What is the benefit of this temporary config dict ?
Couldn't you continue filling in the self.master_config variable ?

M-Mouhcine replied:
That's exactly my comment earlier, on which @cofri responded : 

> I think we introduce this notation of the dict config in config parser to make a clear distinction between the type of the child class and its configuration parameters (i.e. the kwargs to initialize the child class). We did the same for stream handlers and storage handlers.
> 
> This is only done for internal use. In the JSON config file, it's "flattened" as you suggest: the `type` is alongside the other parameters. For example:
> 
> ```json
> "source": {
>         "type": "csv",
>         "path": "path/to/battery_data.csv",
>         "variable": "voltage",
>         "unit": "V",
>         "interpolation": "previous"
>     },
> ```
> 
> We can move `type` inside the config dict to have a single flat dictionary. Both solutions are ok for me.

I lean more towards a single flat dictionary. But it is strictly pedantic as it doesn't heavily impact the code ^^.

cofri committed:
feat: Coordinator instantiates the Master via master_config

The whole config dict for master constructor is given to the
`BaseMaster.create_master(config)`. The parameters about fixed
point initialization are directly added to the master config dict.

There is no need for the `start_master()` method anymore.

jfasquel31450 reviewed (approved):
Ok for me, good job :-)
